jf_agent/git/gitlab_v3_client.py

```diff
@@ -0,0 +1,306 @@
+import gitlab3
+import logging
+import requests
+from jf_agent import agent_logging
+from datetime import datetime, timezone
+
+logger = logging.getLogger(__name__)
+
+
+class MissingSourceProjectException(Exception):
+    pass
+
+
+def log_and_print_request_error(e, action='making request', log_as_exception=False):
+    try:
+        response_code = e.response_code
+    except AttributeError:
+        # if the request error is a retry error, we won't have the code
+        response_code = ''
+
+    error_name = type(e).__name__
+
+    if log_as_exception:
+        agent_logging.log_and_print_error_or_warning(
+            logger, logging.ERROR, msg_args=[error_name, response_code, action, e], error_code=3131,
+        )
+    else:
+        agent_logging.log_and_print_error_or_warning(
+            logger, logging.WARNING, msg_args=[error_name, response_code, action], error_code=3141
+        )
+
+
+class GitLabClient_v3:
+    """
+    __init__(self, server_url, token=None, convert_dates=True, ssl_verify=None, ssl_cert=None)
+
+    Initialize a GitLab connection and optionally supply auth token. 
+    convert_dates can be set to False to disable automatic conversion of date strings to datetime objects. 
+    ssl_verify and ssl_cert are passed to python-requests as the verify and cert arguments, respectively.
+    """
+
+    def __init__(self, server_url, token, convert_dates=True, ssl_verify=None, ssl_cert=None):
+        kwargs = {'token': token, 'convert_dates': convert_dates}
+        if ssl_cert is not None:
+            kwargs['ssl_cert'] = ssl_cert
+        if not ssl_verify:
+            kwargs['ssl_verify'] = False
+        self.server_url = server_url
+        self.agent_args = kwargs
+        self.client = gitlab3.GitLab(server_url, **kwargs)
+        self.version = '3'
```

Comment:
See above -- suggest `self.api_version` instead of `self.version`

Reply:
done